### ISSUE TITLE
Add functionality for querying a Name's NameSpace

### ISSUE THREAD

**Issue by RyanGlScott** [goldfirere/th-desugar#63]
Sometimes, it's not clear in what `NameSpace` a `Name` lives (see, for instance, https://github.com/goldfirere/singletons/pull/230/files#r129394189). What would help is if we had a function of the form:

```
reifyNameSpace :: DsMonad q => Name -> q (Maybe NameSpace)
```

There's a bit of debate to be had about the return type of this function.

* `Maybe NameSpace` is one sensible option, since some `Name`s simply haven't been declared in _any_ `NameSpace` (e.g., `mkName "whargarbl"`, where `whargarbl` isn't a value name).
* On the other hand, there are names which can live in multiple `NameSpace`s (e.g., `Foo` can be a data constructor or a type, and `==` can be a value or a type). Should we instead return `[NameSpace]` to account for this possibility? Or should we adopt the [convention that `reify` uses](https://ghc.haskell.org/trac/ghc/ticket/13961) where if it finds something, it immediately returns that (as a `Just` value)?

**Comment by goldfirere** [comment]
I would favor a more conservative meaning of associated namespace here: a `Name` has a namespace only if it's hard-coded into the name somehow. In other words, I think this function should have type `Name -> Maybe NameSpace`. If the name doesn't have a hard-coded namespace, return `Nothing`. In that case, the caller can then try to `reify`, via various means of reification, and see what they get.

Or perhaps you'd want the function I've described above *and* what you've described.... in which case, yes, perhaps it should return a list.

**Comment by RyanGlScott** [comment]
> I would favor a more conservative meaning of associated namespace here: a `Name` has a namespace only if it's hard-coded into the name somehow. In other words, I think this function should have type `Name -> Maybe NameSpace`. If the name doesn't have a hard-coded namespace, return Nothing. In that case, the caller can then try to `reify`, via various means of reification, and see what they get.

That would fix the problem in https://github.com/goldfirere/singletons/pull/230/files#r129394189, so therefore, that's fine by me.

**Comment by RyanGlScott** [comment]
I take that back - that _wouldn't_ be fine with me. The reason being that a pure version wouldn't work for `singletons`' purposes. I originally tried this simple implementation:

```haskell
nameSpace_maybe :: Name -> Maybe NameSpace
nameSpace_maybe (Name _ nf) =
  case nf of
    NameG ns _ _ -> Just ns
    NameS        -> Nothing
    NameQ{}      -> Nothing
    NameU{}      -> Nothing
    NameL{}      -> Nothing
```

But alas, this doesn't work well for quoted things. Note that in the following example:

```
λ> $([d| data Foo = MkFoo |] >>= \([DataD _ _ _ _ [NormalC n _] _]) -> case n of Name _ nf -> stringE (show nf))
"NameU 6989586621679039213"
```

We get a `NameU` for the quoted data constructor name, which lacks a `NameSpace`. So I don't see a way to make this useful without placing it in `DsMonad`.

**Comment by goldfirere** [comment]
Bah. OK. Do it your way. :)

**Comment by RyanGlScott** [comment]
Fixed in #64.
